internal/wrphandlers/qos/priority_queue.go

```diff
@@ -51,34 +64,69 @@ func (pq *priorityQueue) Enqueue(msg wrp.Message) error {
 	return nil
 }
 
+// trim removes messages with the lowest QualityOfService (taking `prioritizeOldest` into account)
+// until the queue no longer violates `maxQueueSize“.
 func (pq *priorityQueue) trim() {
+	if pq.sizeBytes <= pq.maxQueueBytes {
+		return
+	}
+
+	// Prioritize messages with the lowest QualityOfService such that `pq.Pop()` will return said messages.
+	pq.prioritizeLowestQOS = true
+	defer func() {
+		// Re-prioritize messages with the highest QualityOfService.
+		pq.prioritizeLowestQOS = false
+		// heap.Init() is required since the prioritization was switch to messages with the highest QualityOfService.
+		// The complexity of heap.Init() is O(n) where n = h.Len().
+		heap.Init(pq)
+	}()
+
+	// heap.Init() is required since the prioritization was switch to messages with the lowest QualityOfService.
+	// The complexity of heap.Init() is O(n) where n = h.Len().
+	heap.Init(pq)
```

Comment:
If we end up needing to trim often, then the `O(n)` cost  of `heap.Init()` may start to sting...

I wanted to point that out in case we can't afford the `O(n)` cost  of `heap.Init()` and we would rather trim() random messages instead.